Add test for RTCPeerConnection.addIceCandidate

soareschen opened the pull request:
<!-- Reviewable:start -->

<!-- Reviewable:end -->

soareschen committed:
Add test for RTCPeerConnection.addIceCandidate

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/233150126)

# Firefox (nightly) #
Testing web-platform-tests at revision ec28c76bcbbd45c8d55fa4d6114777d9c7f373b9
Using browser at version BuildID 20170516192431; SourceStamp b8e9b674033bcd1f3a4c59b9d0ee7619c1a17cc5
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5959/webrtc/RTCPeerConnection-addIceCandidate.html">/webrtc/RTCPeerConnection-addIceCandidate.html</a></summary>


|                                                      Subtest                                                      | Results |                                                                                                                                              Messages                                                                                                                                             |
|-------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                   | TIMEOUT |                                                                                                                                                                                                                                                                                                   |
| `Add ICE candidate before setting remote description should reject with InvalidStateError`                        | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add ICE candidate after setting remote description should succeed`                                               | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add ICE candidate with RTCIceCandidate should succeed`                                                           | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Added candidate should be found in remoteDescription`                                                            | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Adding multiple candidates should succeed`                                                                       | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add with empty candidate string (end of candidate) should succeed`                                               | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidCandidateError: Invalid candidate, no ':' ()"`                                                                                                                                                                                      |
| `Add null candidate should reject`                                                                                | FAIL    | `assert_unreached: Should have rejected: undefined Reached unreachable code`                                                                                                                                                                                                                      |
| `Add candidate with both sdpMid and sdpMLineIndex manually set to null should reject with TypeError`              | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid candidate string should reject with OperationError`                                   | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Invalid candidate, no ':' ((Invalid) candidate \r\n string)" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"`                    |
| `Add candidate with invalid candidate string and both sdpMid and sdpMLineIndex null should reject with TypeError` | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with only valid candidate string should reject with TypeError`                                     | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with empty dict should reject with TypeError`                                                      | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with manually filled default values should reject with TypeError`                                  | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with only valid sdpMid should succeed`                                                             | FAIL    | `promise_test: Unhandled rejection with value: object "TypeError: Missing required 'candidate' member of RTCIceCandidateInit."`                                                                                                                                                                   |
| `Add candidate with only valid sdpMLineIndex should succeed`                                                      | FAIL    | `promise_test: Unhandled rejection with value: object "TypeError: Missing required 'candidate' member of RTCIceCandidateInit."`                                                                                                                                                                   |
| `Add candidate with invalid sdpMid should reject with OperationError`                                             | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Mismatch between mid and level - "invalid" is not the mid for level 0; "a1" is" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"` |
| `Add candidate with invalid sdpMLineIndex should reject with OperationError`                                      | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Index 2 out of range" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"`                                                           |
| `Invalid sdpMLineIndex should be ignored if valid sdpMid is provided`                                             | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidCandidateError: Index 2 out of range"`                                                                                                                                                                                              |
| `Add candidate with invalid ufrag should reject with OperationError`                                              | FAIL    | `assert_unreached: Should have rejected: undefined Reached unreachable code`                                                                                                                                                                                                                      |
| `Add candidate with null ufrag should succeed`                                                                    | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with both sdpMid and sdpMLineIndex null should still reject with TypeError after pc is closed`     | TIMEOUT | `Test timed out`                                                                                                                                                                                                                                                                                  |
| `Add valid candidate should never resolve when pc is closed`                                                      | FAIL    | `Peer connection is closed`                                                                                                                                                                                                                                                                       |
| `Add candidate when remote description is null should never resolve when pc is closed`                            | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid candidate string should never resolve when pc is closed`                              | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid sdpMid should never resolve when pc is closed`                                        | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid sdpMLineIndex should never resolve when pc is closed`                                 | PASS    |                                                                                                                                                                                                                                                                                                   |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/233150128)

# Sauce (safari) #
Testing web-platform-tests at revision ec28c76bcbbd45c8d55fa4d6114777d9c7f373b9
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5959/webrtc/RTCPeerConnection-addIceCandidate.html">/webrtc/RTCPeerConnection-addIceCandidate.html</a></summary>


|                                                      Subtest                                                      | Results |                 Messages                 |
|-------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                   | OK      |                                          |
| `Add ICE candidate before setting remote description should reject with InvalidStateError`                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add ICE candidate after setting remote description should succeed`                                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add ICE candidate with RTCIceCandidate should succeed`                                                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Added candidate should be found in remoteDescription`                                                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Adding multiple candidates should succeed`                                                                       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add with empty candidate string (end of candidate) should succeed`                                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add null candidate should reject`                                                                                | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with both sdpMid and sdpMLineIndex manually set to null should reject with TypeError`              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string should reject with OperationError`                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string and both sdpMid and sdpMLineIndex null should reject with TypeError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid candidate string should reject with TypeError`                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with empty dict should reject with TypeError`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with manually filled default values should reject with TypeError`                                  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid sdpMid should succeed`                                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid sdpMLineIndex should succeed`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMid should reject with OperationError`                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMLineIndex should reject with OperationError`                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Invalid sdpMLineIndex should be ignored if valid sdpMid is provided`                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid ufrag should reject with OperationError`                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with null ufrag should succeed`                                                                    | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with both sdpMid and sdpMLineIndex null should still reject with TypeError after pc is closed`     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add valid candidate should never resolve when pc is closed`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate when remote description is null should never resolve when pc is closed`                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string should never resolve when pc is closed`                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMid should never resolve when pc is closed`                                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMLineIndex should never resolve when pc is closed`                                 | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/233150131)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision ec28c76bcbbd45c8d55fa4d6114777d9c7f373b9
Using browser at version 14.14393
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5959/webrtc/RTCPeerConnection-addIceCandidate.html">/webrtc/RTCPeerConnection-addIceCandidate.html</a></summary>


|                                                      Subtest                                                      | Results |              Messages              |
|-------------------------------------------------------------------------------------------------------------------|---------|------------------------------------|
|                                                                                                                   | OK      |                                    |
| `Add ICE candidate before setting remote description should reject with InvalidStateError`                        | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add ICE candidate after setting remote description should succeed`                                               | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add ICE candidate with RTCIceCandidate should succeed`                                                           | FAIL    | `'RTCPeerConnection' is undefined` |
| `Added candidate should be found in remoteDescription`                                                            | FAIL    | `'RTCPeerConnection' is undefined` |
| `Adding multiple candidates should succeed`                                                                       | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add with empty candidate string (end of candidate) should succeed`                                               | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add null candidate should reject`                                                                                | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with both sdpMid and sdpMLineIndex manually set to null should reject with TypeError`              | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid candidate string should reject with OperationError`                                   | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid candidate string and both sdpMid and sdpMLineIndex null should reject with TypeError` | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with only valid candidate string should reject with TypeError`                                     | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with empty dict should reject with TypeError`                                                      | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with manually filled default values should reject with TypeError`                                  | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with only valid sdpMid should succeed`                                                             | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with only valid sdpMLineIndex should succeed`                                                      | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid sdpMid should reject with OperationError`                                             | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid sdpMLineIndex should reject with OperationError`                                      | FAIL    | `'RTCPeerConnection' is undefined` |
| `Invalid sdpMLineIndex should be ignored if valid sdpMid is provided`                                             | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid ufrag should reject with OperationError`                                              | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with null ufrag should succeed`                                                                    | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with both sdpMid and sdpMLineIndex null should still reject with TypeError after pc is closed`     | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add valid candidate should never resolve when pc is closed`                                                      | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate when remote description is null should never resolve when pc is closed`                            | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid candidate string should never resolve when pc is closed`                              | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid sdpMid should never resolve when pc is closed`                                        | FAIL    | `'RTCPeerConnection' is undefined` |
| `Add candidate with invalid sdpMLineIndex should never resolve when pc is closed`                                 | FAIL    | `'RTCPeerConnection' is undefined` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/233150141)

# Chrome (unstable) #
Testing web-platform-tests at revision ec28c76bcbbd45c8d55fa4d6114777d9c7f373b9

rwaldron commented on webrtc/RTCPeerConnection-addIceCandidate.html:
```diff
@@ -0,0 +1,393 @@
+<!doctype html>
+<title>Test RTCPeerConnection.addIceCandidate</title>
```
`RTCPeerConnection.prototype.addIceCandidate`

rwaldron commented:
This is great!

- Minor nit (but not a blocker): the use of semi-colons is really inconsistent. I recommend using them as they are intended to be used (end of expressions, everywhere)
- Only one portion is missing coverage now is 4.6.1: 

> 1. If candidate could not be successfully added the user agent must queue a task that runs the following steps:
>     1. If connection's [[ isClosed]] slot is true, then abort these steps.
>     2. Reject p with a DOMException object whose name attribute has the value OperationError and abort these steps.

I've updated the spec that I'm tracking coverage with https://rwaldron.github.io/webrtc-pc/

soareschen committed:
Add and test additional media stream with different ufrag in SDP

soareschen commented:
My bad. I personally don't use semicolons in my own code so I easily forgot to put it down.

I have updated the code to test SDP containing two media streams with different ufrags. Also add tests to make sure the candidate is added to the correct media stream.

Now all sections in addIceCandidate should be covered.

soareschen committed:
Fix promise_rejects order to make sure addIceCandidate really rejects.

Previously promise_rejects checks on the entire promise. As a result
the rejection may have because of setRemoteDescription rejects or
addIceCandidate throws in the .then() callback.

rwaldron commented:
Excellent!

cc @foolip

rwaldron commented on webrtc/RTCPeerConnection-addIceCandidate.html:
```diff
@@ -0,0 +1,531 @@
+  // WebIDL for addIceCandidate does not allow null argument.
+  // This throws a TypeError because it is checked at WebIDL level.
+  // null can be accidentally passed from onicecandidate event handler
+  // when null is used to indicate end of candidate
```
Great test

soareschen replied:
Actually that is wrong. I later remembered WebIDL explicitly states that if the return type of the method is a promise, it converts any error thrown during checking to promise rejection. I updated the code but forgot to remove the comment.

https://heycam.github.io/webidl/#es-operations

foolip commented on webrtc/RTCPeerConnection-addIceCandidate.html:
```diff
@@ -0,0 +1,531 @@
+      }));
+      pc.close();
+    });
+  }, 'Add candidate with invalid candidate string should never resolve when pc is closed');
```
This is a duplicate test name, doesn't it cause a harness error?

soareschen replied:
Hmm I must have made one extra copy when duplicating and editing the tests. Will remove.

I think testharness doesn't report it because it report the timeout error instead for _Add candidate with both sdpMid and sdpMLineIndex null should still reject with TypeError after pc is closed_.

soareschen committed:
Fix incorrect comment description and remove duplicate test

soareschen committed:
Add TODO note on more tests to be added

soareschen commented:
There are still some more test need to be added, but this PR can be squashed and merged first.

I have added a TODO comment to note that the following tests are still needed:

- Set pc with both current and pending remote descriptions with different ufrags, check that addIceCandidate with specific ufrag adds candidate to the correct remote description. This cannot be tested for now because there is a bug in the spec: w3c/webrtc-pc#1227

- Call with candidate string containing partial malformed syntax, i.e. malformed IP. Some browsers like Firefox may ignore the syntax error and add it to the SDP regardless.

rwaldron commented on webrtc/RTCPeerConnection-addIceCandidate.html:
```diff
@@ -514,4 +514,11 @@
+  //   remote description.
+  // - Call with candidate string containing partial malformed syntax, i.e. malformed IP.
+  //   Some browsers may ignore the syntax error and add it to the SDP regardless.
+
```
👍

rwaldron commented:
LGTM
